include/ensmallen_bits/moead/moead_impl.hpp

```diff
@@ -0,0 +1,453 @@
+  {
+    if (lowerBound[geneIdx] >= upperBound[geneIdx])
+    {
+      std::ostringstream ss;
+      ss << "MOEAD::Optimize():" << "the lowerBound value: " << lowerBound[geneIdx]
+         << " is greater than upperBound value: " << upperBound[geneIdx]
+         << " at index: " << geneIdx << std::endl;
+
+      throw std::logic_error(ss.str());
+    }
+  }
+
+  bool terminate = false;
+
+  arma::uvec shuffle;
+  // The weight matrix. Each vector represents a decomposition subproblem.
+  arma::Mat<ElemType> weights(numObjectives, populationSize, arma::fill::randu);
+  weights += 1E-10; // Numerical stability
+
+  // 1.2 Storing the indices of nearest neighbors of each weight vector.
+  arma::umat neighborIndices(neighborSize, populationSize);
+  for (size_t i = 0; i < populationSize; ++i)
+  {
+    // Cache the distance between weights(i) and other weights.
+    arma::rowvec distances(populationSize);
+    distances =
+        arma::sqrt(arma::sum(arma::square(weights.col(i) - weights.each_col())));
+    arma::uvec sortedIndices = arma::stable_sort_index(distances);
+    // Ignore distance from self
+    neighborIndices.col(i) = sortedIndices(arma::span(1, neighborSize));
+  }
+
+  // 1.2 Random generation of the initial population.
+  std::vector<MatType> population(populationSize);
+  for (size_t i = 0; i < populationSize; ++i)
+  {
+	  population[i] =
+		    arma::randu<MatType>(iterate.n_rows, iterate.n_cols) - 0.5 + iterate;
+
+	  for (size_t geneIdx = 0; geneIdx < numVariables; ++geneIdx)
+	  {
+		  if (population[i][geneIdx] < lowerBound[geneIdx])
+			  population[i][geneIdx] = lowerBound[geneIdx];
+		  else if (population[i][geneIdx] > upperBound[geneIdx])
+			  population[i][geneIdx] = upperBound[geneIdx];
+	  }
+  }
+
+  arma::mat populationFitness(numObjectives, populationSize);
+  EvaluateObjectives(population, objectives, populationFitness);
+
+  // 1.3 Initialize the ideal point z.
+  arma::vec idealPoint(numObjectives);
+  idealPoint.fill(std::numeric_limits<double>::max());
+  for (size_t i = 0; i < numObjectives; ++i)
+  {
+    for (size_t j = 0; j < populationSize; ++j)
+    {
+      idealPoint(i) = std::min(idealPoint(i), populationFitness(i, j));
+    }
+  }
+
+  terminate |= Callback::BeginOptimization(*this, objectives, iterate, callbacks...);
+
+  // 2 The main loop.
+  for (size_t g = 0; g < maxGenerations; ++g)
+  {
+    shuffle = arma::shuffle(
+        arma::linspace<arma::uvec>(0, populationSize - 1, populationSize));
+    for (size_t i : shuffle)
+    {
+      terminate |= Callback::StepTaken(*this, objectives, iterate, callbacks...);
+
+      Info << "MOEA/D-DE initialized successfully. Optimization started." << std::endl;
+
+      // 2.1 Randomly select two indices in neighborIndices(i) and use them
+      // to make a child.
+      size_t r1, r2, r3;
+      r1 = i;
+      // Randomly choose to sample from the population or the neighbors.
+      bool sampleNeighbor = ( arma::randu() < neighborProb );
```

Comment:
```suggestion
      bool sampleNeighbor = arma::randu() < neighborProb;
```

No need for the extra `()`.